indy_ledger_response_parser/src/domain/response.rs

```diff
@@ -0,0 +1,129 @@
+use indy_api_types::{
+    errors::{err_msg, IndyErrorKind},
+    IndyError,
+};
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct Response {
+    pub req_id: u64,
+    pub reason: String,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+pub enum Reply<T> {
+    ReplyV0(ReplyV0<T>),
+    ReplyV1(ReplyV1<T>),
+}
+
+impl<T> Reply<T> {
+    pub fn result(self) -> T {
+        match self {
+            Reply::ReplyV0(reply) => reply.result,
+            Reply::ReplyV1(mut reply) => reply.data.result.remove(0).result,
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ReplyV0<T> {
+    pub result: T,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ReplyV1<T> {
+    pub data: ReplyDataV1<T>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ReplyDataV1<T> {
+    pub result: Vec<ReplyV0<T>>,
```

Comment:
While working on the suggestion to avoid double deserialization, I was wondering why is this a `Vec`, since it looks like in the `result()` method the first element is removed and the others, if any, are lost. If there's only one element to be expected, how about making it an array?
The element can then be extracted with `result.into_iter().next().unwrap()` which is sound since you cannot instantiate an unfilled array.

```suggestion
    pub result: [ReplyV0<T>; 1],
```

Reply:
If there is always only one element to be expected, I don't see the point of making it any kind of array in the first place. I don't know why it is a `Vec` - perhaps this is to support some sort of batch transactions in the future...? Then the current implementation of using only the first element would be a temporary workaround, and so I would probably prefer to not accommodate the type to the current use case. But again, as I said, I don't know.

Reply:
Why not deal with the idea of batch transactions if/when it will be a thing? Whether the workaround is temporary or not, I'd say it could at least be a better workaround. But feel free to ignore this and resolve the conversation if it's not important for you.

Reply:
Alright, changed the `Vec` to an array.